src/Features/Core/Portable/CodeRefactorings/CodeRefactoringService.cs

```diff
@@ -163,5 +176,107 @@ private IEnumerable<CodeRefactoringProvider> GetProviders(Document document)
 
             return null;
         }
+
+        private ImmutableArray<CodeRefactoringProvider> GetProjectRefactorings(Project project)
+        {
+            // TODO (https://github.com/dotnet/roslyn/issues/4932): Don't restrict refactorings in Interactive
+            return project.Solution.Workspace.Kind == WorkspaceKind.Interactive
+                ? ImmutableArray<CodeRefactoringProvider>.Empty
+                : _projectRefactoringsMap.GetValue(project.AnalyzerReferences, pId => new StrongBox<ImmutableArray<CodeRefactoringProvider>>(ComputeProjectRefactorings(project))).Value;
+        }
+
+        private ImmutableArray<CodeRefactoringProvider> ComputeProjectRefactorings(Project project)
+        {
+            var extensionManager = project.Solution.Workspace.Services.GetService<IExtensionManager>();
+            ImmutableArray<CodeRefactoringProvider>.Builder builder = null;
+            foreach (var reference in project.AnalyzerReferences)
+            {
+                var projectCodeRefactoringProvider = _analyzerReferenceToRefactoringsMap.GetValue(reference, _createProjectCodeRefactoringsProvider);
+                foreach (var refactoring in projectCodeRefactoringProvider.GetRefactorings(project.Language))
+                {
+                    builder ??= ImmutableArray.CreateBuilder<CodeRefactoringProvider>();
+                    builder.Add(refactoring);
+                }
+            }
+
+            if (builder is null)
+            {
+                return ImmutableArray<CodeRefactoringProvider>.Empty;
+            }
+
+            return builder.ToImmutable();
+        }
+
+        private class ProjectCodeRefactoringProvider
+        {
+            private readonly AnalyzerReference _reference;
+            private ImmutableDictionary<string, ImmutableArray<CodeRefactoringProvider>> _refactoringsPerLanguage;
+
+            public ProjectCodeRefactoringProvider(AnalyzerReference reference)
+            {
+                _reference = reference;
+                _refactoringsPerLanguage = ImmutableDictionary<string, ImmutableArray<CodeRefactoringProvider>>.Empty;
+            }
+
+            public ImmutableArray<CodeRefactoringProvider> GetRefactorings(string language)
+            {
+                return ImmutableInterlocked.GetOrAdd(ref _refactoringsPerLanguage, language, CreateRefactorings);
```

Comment:
should there be a Delegate pre-created for 'CreateRefactorings' to void the alloc?

Reply:
this is lazily computed.  but is there any chance we *won't* need these *once* we've made the ProjectCodeRefactoringProvider.  If we'll always need the values, perhaps we should just compute up front.

Reply:
This now uses a non-capturing lambda